Fixed apart() producing wrong answer

citizen-seven opened the pull request:
apart() produced a wrong result when given multivariate polynomial with fractional coefficients in a result. Division was done correctly, but the domain wasn't properly converted to a field. Now the default parameter `field=True` is set explicitly before conversion to polynomial.

Previously:
`In [1]: apart(x/2, y)`
`Out[1]: 0`

Now:
`In [1]: apart(x/2, y)`
`Out[1]:`
x
─
2

This fixes #12177, #9123.

citizen-seven committed:
Fixes #12177, #9123: "Wrong result with apart",  "apart drops term".

The problem: division was made within the ZZ ring, not the QQ field.
Setting manually the domain to be a field while constructing polynomial solves the problem.

Previously:
In [1]: apart(x/2, y)
Out[1]: 0

Now:
In [1]: apart(x/2, y)
Out[1]:
x
─
2

kshitij10496 commented:
@citizen-seven Good work here !
Kindly add suitable tests corresponding to the changes you have made. 
Also, I would suggest you to change your commit message so that it briefly explains the changes you have made.

jksuom commented:
It seems to me that the reason for the error is the failure of the division `poly, P = P.div(Q, auto=True)` in case `P = Poly(x + y, x)` and `Q = Poly(2*x - y, x)`, for example.

When that is fixed it should not be necessary to add `field=True` (though that is probably not harmful either).

citizen-seven commented:
@jksuom Well, `div()` works as it is specified in the [documentation](http://docs.sympy.org/latest/modules/polys/basics.html#division), however the `div()` method of the Poly class doesn't has a keyword argument `domain`, what needs to be fixed, I guess.

jksuom commented:
> div() works as it is specified in the documentation

I don't think `div()` works correctly if `P.div(Q)` returns a pair of zero polynomials when `P` and `Q` are nonzero.

citizen-seven commented:
@jksuom `div()` public method explicitly uses `div()` method of the Poly class, so we should specify the domain for it to work as expected:
`In[]: P = Poly(x + y, x, domain=QQ[y])`
`Q = Poly(2*x - y, x, domain=QQ[y])`
`poly = P.div(Q)`
`poly`
`Out[]: (Poly(1/2, x, domain='QQ[y]'), Poly(3*y/2, x, domain='QQ[y]'))`

jksuom commented:
It is expected that the methods would automatically find default domains without the user having to spcecify them.
```
In [5]: Poly(x + y, x).div(Poly(2*x - y, x))
Out[5]: (Poly(1/2, x, domain='ZZ(y)'), Poly(3*y/2, x, domain='ZZ(y)'))
```

citizen-seven commented:
@jksuom So, I guess, the conversion of ZZ[y] to ZZ(y) is lost somewhere.

citizen-seven commented:
@jksuom After division do we need to try to convert ZZ(y) to ZZ[y] back? I've just found out that the reason the following code works is that the 2*a (the denominator) is not ground.
```
 In[]: Poly(a+y, y).div(Poly(2*a, y)
 Out[]: Out[2]: (Poly(1/(2*a)*y + 1/2, y, domain='ZZ(a)'), Poly(0, y, domain='ZZ(a)'))
```
However, the next division isn't done correctly because of denom(1/2) = 2 being ground.
```
In[]: Poly(x+a, x).div(Poly(2*x+a, x))
Out[]: (Poly(0, x, domain='ZZ[a]'), Poly(0, x, domain='ZZ[a]'))
```
In general case, we try to retract the lowest domain. In tests there is `Poly(x**2).div(Poly(x))`, which is done in QQ and then converted back to a ZZ.

The problem is in the `K0.numer(a)/denom`, which returns zero and further conversion has no sense:
```
if denom.is_ground:     
    return K1.from_PolynomialRing(K0.numer(a)/denom, K0.field.ring.to_domain())
```
What is going to be the best possible solution?

jksuom commented:
I also think that the bug is in the conversion from fractions to polynomials. It seems that the code was originally written assuming that the coefficient domain would always be a field, in which case `denom.is_ground` (i.e,. a constant polynomial) would suffice. When there are noninvertible nonzero constants, we need a better test. I would suggest a solution of the following type:
```
        quotient, remainder = K0.numer(a).div(K0.denom(a))

        if remainder.is_zero:
            return K1.from_PolynomialRing(quotient, K0.field.ring.to_domain())
        else:
            return None
```

citizen-seven committed:
Fixed division for polynomials with parameters & tests

The division of polynomials was done with conversion to the FractionField and, if it's possible, with reverse conversion to a Ring.

The condition in from_FractionField() method of a PolynomialRing class was wrong (the division wasn't using div() method of it's class) and it lead to 0 result in some cases, e.g.:

Problem:
In[]: Poly(x+y, x).div(Poly(2*x-y, x))
Out[]: (Poly(0, x, domain='ZZ[y]'), Poly(0, x, domain='ZZ[y]'))

Now:
In[]: Poly(x+y, x).div(Poly(2*x-y, x))
Out[]: (Poly(1/2, x, domain='ZZ(y)'), Poly(3*y/2, x, domain='ZZ(y)'))

Moreover, this lead to the incorrect results in apart(), so the previous solution has no sense.

Proper tests were added.

jksuom commented on sympy/polys/tests/test_polytools.py:
```diff
@@ -1700,6 +1700,14 @@ def test_div():
+    f, g = Poly(x+y, x), Poly(2*x+y, x)
+    q, r = f.div(g)
+    assert q.get_domain().is_Frac and r.get_domain().is_Frac
+    r = f.rem(g)
```
Maybe this rem test and the quo test below are not necessary since they are the same as the results `q, r` from div above.

citizen-seven committed:
Tests update

Quo() and rem() methods use the same convertation as the div, so the result after division is the same. That's why some tests are unnecessary.

jksuom commented:
Thanks! This looks good.
